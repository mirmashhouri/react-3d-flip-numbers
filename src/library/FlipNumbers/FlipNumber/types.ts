import type { FC } from 'react';

export type IFlipNumber = FC<{
  duration: number;
  activeNumber: number;
  delay: number;
  direction: 1 | -1;
  height: number;
  hasShadow: boolean;
  className?: string;
}>;

export interface FlipNumberState {
  degree: number;
  rotateCounter: number;
}
