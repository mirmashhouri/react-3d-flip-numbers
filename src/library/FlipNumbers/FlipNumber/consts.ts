export const REVOLUTION_DEGREES = 360;
export const ROTATE_DEGREE_PER_NUMBER = 36;
export const NUMBER_ARRAY = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
