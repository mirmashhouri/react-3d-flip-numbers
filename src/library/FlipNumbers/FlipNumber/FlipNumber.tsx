import React, { useEffect, useState } from 'react';
import { NUMBER_ARRAY, ROTATE_DEGREE_PER_NUMBER, REVOLUTION_DEGREES } from './consts';
import { IFlipNumber } from './types';

export const FlipNumber: IFlipNumber = ({
  className = '',
  activeNumber,
  duration,
  delay,
  direction,
  height,
  hasShadow,
}) => {
  const [rotateCounter, setRotateCounter] = useState(0);
  const [degree, setDegree] = useState(0);
  const [visibleActive, setVisibleActive] = useState(true);
  const [prevAnimateDegree, setPrevAnimateDegree] = useState(0);
  const [fixedActiveNum, setFixedActiveNum] = useState(activeNumber);
  const [numberDirection, setNumberDirection] = useState<1 | -1>(direction);

  const calculateDegrees = (newActiveNumber: number, newDuration: number, newDelay: number) => {
    setVisibleActive(false);
    const animateDegree =
      NUMBER_ARRAY.findIndex((v) => v === newActiveNumber) * ROTATE_DEGREE_PER_NUMBER;

    const amountDegree = rotateCounter * REVOLUTION_DEGREES;
    let newDegree = numberDirection * amountDegree - animateDegree;

    if (
      (numberDirection === -1 && newDegree > degree) ||
      (numberDirection === 1 && newDegree < degree)
    ) {
      const newRotateCounter = rotateCounter + 1;
      const newAmountDegree = newRotateCounter * REVOLUTION_DEGREES;
      setRotateCounter(newRotateCounter);
      newDegree = numberDirection * newAmountDegree - animateDegree;
    }

    setDegree(newDegree);
    setPrevAnimateDegree(animateDegree);

    setTimeout(() => {
      setVisibleActive(true);
    }, newDuration + newDelay);
  };

  const resetRotateCounter = (newDdirection: 1 | -1) => {
    setNumberDirection(newDdirection);
    setRotateCounter(0);
    const redirectDegree = -prevAnimateDegree;
    setDegree(redirectDegree);
  };

  useEffect(() => {
    resetRotateCounter(direction);
  }, [direction]);

  useEffect(() => {
    if (fixedActiveNum !== activeNumber) {
      calculateDegrees(activeNumber, duration, delay);
      setFixedActiveNum(activeNumber);
    }
  }, [activeNumber, fixedActiveNum]);

  useEffect(() => {
    calculateDegrees(activeNumber, duration, delay);
    setFixedActiveNum(activeNumber);
  }, []);

  const translateZ = height * 1;

  return (
    <div
      style={ { height } }
      className={ `flip-number-container ${ hasShadow && !visibleActive ? 'flip-shadow' : ''} ${className}` }
    >
      <div
        className="flip-number"
        style={ {
          transform: `rotateX(${degree}deg)`,
          transition: `transform ${duration}ms ${delay}ms ease-out`,
        } }
      >
        { NUMBER_ARRAY.map((num, i) => (
          <div
            className={ `flip-number-fake ${visibleActive ? 'invisible' : ''}` }
            style={ {
              height,
              transform: `rotateX(${ROTATE_DEGREE_PER_NUMBER * i}deg) translateZ(${translateZ}px) ${
                num === activeNumber ? 'scale3d(1, 1, 1)' : 'scale3d(1, .8, .8)'
              }`,
            } }
            key={ `${ROTATE_DEGREE_PER_NUMBER * i}` }
          >
            { num }
          </div>
        )) }
      </div>

      <div
        className={ `flip-number-active ${visibleActive ? '' : 'invisible'}` }
        style={ { height } }
      >
        { activeNumber }
      </div>
    </div>
  );
};
