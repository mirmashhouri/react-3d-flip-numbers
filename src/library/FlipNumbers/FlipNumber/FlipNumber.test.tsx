import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import { FlipNumber } from './FlipNumber';

const component =
<FlipNumber
  activeNumber={ 4 }
  duration= { 500 }
  delay= { 0 }
  direction= { -1 }
  height={ 50 }
  hasShadow= { false }
/>;

describe('FlipNumber component', () => {
  test('renders FlipNumber component', () => {
    const { container } = render(component);
    expect(container).toBeInTheDocument();
  });

  test('renders fake numbers', () => {
    const fakeClassName = 'flip-number-fake';
    const fakeEl = render(component).container.getElementsByClassName(fakeClassName);
    expect(fakeEl).toHaveLength(10);
  });

  test('renders flip-number-active', () => {
    const activeClassName = 'flip-number-active';
    const activeEl = render(component).container.getElementsByClassName(activeClassName);
    expect(activeEl).toHaveLength(1);
    expect(activeEl[0]).toHaveTextContent(component.props.activeNumber);
  });

  test('make snapshot', () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
