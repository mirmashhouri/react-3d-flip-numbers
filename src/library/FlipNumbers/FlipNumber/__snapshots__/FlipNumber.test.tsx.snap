// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipNumber component make snapshot 1`] = `
<div>
  <div
    class="d-table text-start overflow-hidden position-relative z-index-main-view "
    style="height: 50px;"
  >
    <div
      class="flip-number position-absolute h-100"
      style="transform: rotateX(-144deg); transition: transform 500ms 0ms ease-in-out;"
    >
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
      >
        0
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
      >
        1
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
      >
        2
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
      >
        3
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
      >
        4
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
      >
        5
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
      >
        6
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
      >
        7
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
      >
        8
      </div>
      <div
        class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
        style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
      >
        9
      </div>
    </div>
    <div
      class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
      style="height: 50px;"
    >
      4
    </div>
  </div>
</div>
`;
